packages/bug-detectors/internal/eval.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2023 Code Intelligence GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+	getJazzerJsGlobal,
+	guideTowardsContainment,
+	reportAndThrowFinding,
+	setJazzerJsGlobal,
+} from "@jazzer.js/core";
+import { callSiteId } from "@jazzer.js/hooking";
+import * as vm from "vm";
+
+const targetString = "jaz_zer";
+
+const originalEval = eval;
+const originalFunction = Function;
+
+globalThis.eval = hookEval(originalEval);
+// @ts-ignore
+globalThis.Function = hookFunction(originalFunction);
+
+if (getJazzerJsGlobal("vmContext")) {
+	const vmContext = getJazzerJsGlobal("vmContext") as vm.Context;
+	// The VM should use eval from its own context, and not the global eval.
+	vmContext.eval = hookEval(vm.runInContext("eval", vmContext));
+	vmContext.Function = hookFunction(vm.runInContext("Function", vmContext));
+}
+
+function hookEval(originalEval: typeof eval) {
+	return function hookedEval(code: string) {
+		if (code.includes(targetString)) {
+			reportAndThrowFinding(
+				`Arbitrary code execution using eval:\n        '${code}'`,
+			);
+		}
+
+		// Since we do not hook eval using the hooking framework, we have to recompute the
+		// call site ID on every call to eval. This shouldn't be an issue, because eval is
+		// considered evil and should not be called too often, or even better -- not at all!
+		guideTowardsContainment(code, targetString, callSiteId());
+
+		return originalEval(code);
+	};
+}
+
+function hookFunction(originalFunction: (...args: string[]) => unknown) {
+	return function hookedFunction(...args: string[]) {
+		if (args.length > 0) {
+			const functionBody = args[args.length - 1];
+			if (typeof functionBody === "string") {
```

Comment:
Suppress warning for this statement. Calling code could pass in something different from a string so the check in itself is reasonable.